File: src/main/java/pl/training/workshop/shared/Money.java
package pl.training.workshop.shared;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

/**
 * Kwota w złotych, zawsze ze skalą 2 i zaokrągleniem HALF_UP.
 * Wspólny, nierefaktoryzowany typ wartości warsztatu CineLegacy.
 */
public record Money(BigDecimal amount) implements Comparable<Money> {
    public static final Money ZERO = Money.of("0.00");

    public Money {
        Objects.requireNonNull(amount, "amount");
        amount = amount.setScale(2, RoundingMode.HALF_UP);
    }

    public static Money of(String amount) {
        return new Money(new BigDecimal(amount));
    }

    public static Money of(long amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    public Money plus(Money other) {
        return new Money(amount.add(other.amount));
    }

    public Money minus(Money other) {
        return new Money(amount.subtract(other.amount));
    }

    public Money times(int factor) {
        return new Money(amount.multiply(BigDecimal.valueOf(factor)));
    }

    /** Procent kwoty, np. {@code percent(25)} to 25% tej kwoty. */
    public Money percent(int percent) {
        return new Money(amount
                .multiply(BigDecimal.valueOf(percent))
                .divide(BigDecimal.valueOf(100), 2, RoundingMode.HALF_UP));
    }

    public Money max(Money other) {
        return compareTo(other) >= 0 ? this : other;
    }

    public boolean isGreaterThan(Money other) {
        return compareTo(other) > 0;
    }

    @Override
    public int compareTo(Money other) {
        return amount.compareTo(other.amount);
    }

    @Override
    public String toString() {
        return amount.toPlainString();
    }
}

File: typescript/src/workshop/shared/Money.ts
import { Decimal } from 'decimal.js';

import { requireNonNull } from '../../shared/requireNonNull.js';

// Kwota w złotych, zawsze ze skalą 2 i zaokrągleniem HALF_UP.
// Wspólny, nierefaktoryzowany typ wartości warsztatu CineLegacy.
export class Money {
  static readonly ZERO = Money.of('0.00');

  readonly amount: Decimal;

  constructor(amount: Decimal) {
    this.amount = requireNonNull(amount, 'amount').toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
  }

  static of(amount: string | number): Money {
    return new Money(new Decimal(amount));
  }

  plus(other: Money): Money {
    return new Money(this.amount.plus(other.amount));
  }

  minus(other: Money): Money {
    return new Money(this.amount.minus(other.amount));
  }

  times(factor: number): Money {
    return new Money(this.amount.times(factor));
  }

  // Procent kwoty, np. percent(25) to 25% tej kwoty.
  percent(percent: number): Money {
    return new Money(this.amount.times(percent).dividedBy(100));
  }

  max(other: Money): Money {
    return this.compareTo(other) >= 0 ? this : other;
  }

  isGreaterThan(other: Money): boolean {
    return this.compareTo(other) > 0;
  }

  compareTo(other: Money): number {
    return this.amount.comparedTo(other.amount);
  }

  equals(other: unknown): boolean {
    return other instanceof Money && this.amount.equals(other.amount);
  }

  toString(): string {
    return this.amount.toFixed(2);
  }
}
